Add AllowTopicCreation

zorkian opened the pull request:
This adds a broker configuration option to enable automatic topic
creation. This defaults to false.

This is useful in clusters where you have a dumb proxy that is
forwarding messages into Kafka which has topic creation enabled. Note
that this is not the fastest of operations, it is effectively a metadata
miss, so requires at least two server roundtrips.

zorkian commented:
cc @DrTall @dcrwatson

optiopay-backend-helper commented:
Pull request seem to be stale, assigning @alpe as the responsible developer.

husio commented:
Code in this branch is panicking. 

```sh
% go run testcreate.go -kafka `kafka_addr` -topic "test-topic-001"
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x18 pc=0x46da80]

goroutine 1 [running]:
github.com/optiopay/kafka.(*connection).Offset(0x0, 0xc208100340, 0x1, 0x0, 0x0)
        /home/piotr/optiopay/src/github.com/optiopay/kafka/connection.go:279 +0x50
github.com/optiopay/kafka.(*Broker).offset(0xc20807e000, 0x7ffc96e5f47b, 0xe, 0x0, 0xfffffffffffffffe, 0x0, 0x0, 0x0)
        /home/piotr/optiopay/src/github.com/optiopay/kafka/broker.go:582 +0x2ea
github.com/optiopay/kafka.(*Broker).OffsetEarliest(0xc20807e000, 0x7ffc96e5f47b, 0xe, 0x0, 0x0, 0x0, 0x0)
        /home/piotr/optiopay/src/github.com/optiopay/kafka/broker.go:630 +0x5f
github.com/optiopay/kafka.(*Broker).Consumer(0xc20807e000, 0x7ffc96e5f47b, 0xe, 0x0, 0x2faf080, 0xffffffffffffffff, 0x2faf080, 0xa, 0x1dcd6500, 0x1e848000000001, ...)
        /home/piotr/optiopay/src/github.com/optiopay/kafka/broker.go:918 +0x158
main.main()
        /home/piotr/optiopay/src/github.com/optiopay/testcreate.go:28 +0x5c4

goroutine 101 [runnable]:
net.(*pollDesc).Wait(0xc2080ae140, 0x72, 0x0, 0x0)
        /usr/lib/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc2080ae140, 0x0, 0x0)
        /usr/lib/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).Read(0xc2080ae0e0, 0xc2080ad000, 0x1000, 0x1000, 0x0, 0x7fa2a892da40, 0xc20811a098)
        /usr/lib/go/src/net/fd_unix.go:242 +0x40f
net.(*conn).Read(0xc2080a2008, 0xc2080ad000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        /usr/lib/go/src/net/net.go:121 +0xdc
bufio.(*Reader).fill(0xc2080f0180)
        /usr/lib/go/src/bufio/bufio.go:97 +0x1ce
bufio.(*Reader).Read(0xc2080f0180, 0xc2080dc800, 0x4, 0x400, 0x400, 0x0, 0x0)
        /usr/lib/go/src/bufio/bufio.go:174 +0x26c
io.ReadAtLeast(0x7fa2a892ed28, 0xc2080f0180, 0xc2080dc800, 0x4, 0x400, 0x4, 0x0, 0x0, 0x0)
        /usr/lib/go/src/io/io.go:298 +0xf1
io.ReadFull(0x7fa2a892ed28, 0xc2080f0180, 0xc2080dc800, 0x4, 0x400, 0x544000, 0x0, 0x0)
        /usr/lib/go/src/io/io.go:316 +0x6d
github.com/optiopay/kafka/proto.(*decoder).DecodeInt32(0xc208100240, 0xc2080dc800)
        /home/piotr/optiopay/src/github.com/optiopay/kafka/proto/serialization.go:64 +0x80
github.com/optiopay/kafka/proto.ReadResp(0x7fa2a892ed28, 0xc2080f0180, 0xc20809bf28, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/piotr/optiopay/src/github.com/optiopay/kafka/proto/messages.go:86 +0x16f
github.com/optiopay/kafka.(*connection).readRespLoop(0xc2080bd720)
        /home/piotr/optiopay/src/github.com/optiopay/kafka/connection.go:83 +0x100
created by github.com/optiopay/kafka.newTCPConnection
        /home/piotr/optiopay/src/github.com/optiopay/kafka/connection.go:46 +0x2d7
exit status 2


% cat testcreate.go                                                                                                          :(
package main

import (
        "flag"
        "fmt"
        "log"

        "github.com/optiopay/kafka"
)

const kafkaAddrs = "localhost:9092"

func main() {
        var (
                kafkaAddrFl = flag.String("kafka", "", "Kafka address")
                topicFl     = flag.String("topic", "test-topic", "Topic to create")
        )
        flag.Parse()

        conf := kafka.NewBrokerConf(fmt.Sprintf("test-create:%s", *topicFl))
        conf.AllowTopicCreation = true
        broker, err := kafka.Dial([]string{*kafkaAddrFl}, conf)
        if err != nil {
                log.Fatalf("cannot connect to %s: %s", *kafkaAddrFl, err)
        }
        defer broker.Close()

        if _, err := broker.Consumer(kafka.NewConsumerConf(*topicFl, 0)); err != nil {
                log.Fatalf("cannot create consumer: %s", err)
        }
}

```

zorkian commented:
Ok so the problem here was that in the case where the fetch to the server did not result in a topic creation, we would exhaust retries and return `nil, nil` causing the connection logic to fail (rightly so).

We now proactively set an error assuming that the creation will fail, and if we run out of retries having never seen the topic, we'll just return that. It's probably the best failure mode we can hope for in this particular situation.

I also added tests for the two code paths, and refactored a bit of the metadata tester code to make it a bit more flexible and support topic creation.

husio commented on broker.go:
```diff
@@ -361,11 +382,22 @@ func (b *Broker) muLeaderConnection(topic string, partition int32) (conn *connec
+				if b.conf.AllowTopicCreation {
+					_, err := b.fetchMetadata([]string{topic})
+					if err != nil {
+						b.conf.Logger.Info("failed to fetch metadata for new topic %s: %s",
```
Logger does not allow formatting. First parameter should be description and than any number of key-value pairs. In this case, it should be 
```go
b.conf.Logger.Info("failed to fetch metadata for new topic", 
    "topic", topic,
     "err", err)
```

husio commented on broker.go:
```diff
@@ -248,9 +256,13 @@ func (b *Broker) muRefreshMetadata() error {
 // Because it's using metadata information to find node connections it's not
 // thread safe and using it require locking.
-func (b *Broker) fetchMetadata() (*proto.MetadataResp, error) {
+func (b *Broker) fetchMetadata(topics []string) (*proto.MetadataResp, error) {
```
What do you think about changing this function notation to accept `(topic ...string)`? I need to check if empty slice will work the same way as `nil`, but I think it's nicer to use with `...`

zorkian commented:
Oh yeah, I think `...topic` is better. I usually forget you can do that. Cleaned up, tests still pass.

Passing `nil` for a slice just ends up with an empty slice (it's the zero value), so passing an empty slice should be functionally equivalent. (I had to research this.)

husio commented:
That looks really good now! Tomorrow I will test it locally and merge it.

husio commented:
Are you sure it's working? I'm trying to both consume and produce to non existing topic with this code and it fails, because it's trying to fetch metadata for all topics, not any in particular:

```go
package main

import (
        "flag"
        "fmt"
        "os"

        "github.com/husio/log"
        "github.com/optiopay/kafka"
        "github.com/optiopay/kafka/proto"
)

func main() {
        var (
                kafkaAddrFl = flag.String("kafka", "", "Kafka address")
                topicFl     = flag.String("topic", "test-topic", "Topic to create")
        )
        flag.Parse()

        conf := kafka.NewBrokerConf(fmt.Sprintf("test-create:%s", *topicFl))
        conf.Logger = &logger{log.NewJSONLogger(os.Stdout)}
        conf.AllowTopicCreation = true
        broker, err := kafka.Dial([]string{*kafkaAddrFl}, conf)
        if err != nil {
                log.Fatal("cannot connect",
                        "address", *kafkaAddrFl,
                        "err", err.Error())
        }
        defer broker.Close()

        /*
                if _, err := broker.Consumer(kafka.NewConsumerConf(*topicFl, 0)); err != nil {
                        log.Error("cannot create consumer", "err", err.Error())
                }
        */
        msg := proto.Message{Value: []byte("foo")}
        if _, err := broker.Producer(kafka.NewProducerConf()).Produce(*topicFl, 0, &msg); err != nil {
                log.Error("cannot produce", "err", err.Error())
        }
}

type logger struct {
        log log.Logger
}

func (lg *logger) Debug(msg string, keyvals ...interface{}) {
        lg.log.Debug(msg, tostring(keyvals)...)
}

func (lg *logger) Info(msg string, keyvals ...interface{}) {
        lg.log.Info(msg, tostring(keyvals)...)
}

func (lg *logger) Warn(msg string, keyvals ...interface{}) {
        lg.log.Warn(msg, tostring(keyvals)...)
}

func (lg *logger) Error(msg string, keyvals ...interface{}) {
        lg.log.Error(msg, tostring(keyvals)...)
}

func (lg *logger) Crit(msg string, keyvals ...interface{}) {
        lg.log.Crit(msg, tostring(keyvals)...)
}

func tostring(keyvals ...interface{}) []string {
        var pairs []string
        for _, v := range keyvals {
                if s, ok := v.(string); ok {
                        pairs = append(pairs, s)
                } else {
                        pairs = append(pairs, fmt.Sprint(v))
                }
        }
        return pairs
}
```

zorkian commented:
Hrmmmm. It's working on production here. But I can't swear what we have running in prod is the exact same as what's in this PR.

This week we've got basically a company offsite so I don't have much time, but I will try to get back to this over the weekend or early next week. Apologies for the delay.

zorkian commented:
Are you sure you have your Kafka cluster configured to allow topic creation? We're using this code in production and it's working just fine for us with 0.8.2. It creates the topics and publishes, etc.

husio commented:
You're right, I was testing it with different docker image than I updated config for.
